Bottom: 99ac641d67d5f291fccd17eb9905fe57c3449c84
Top:    cb2f4229b46900ff0f7a4e1e12b023899ea404bc
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-12-04 19:16:18 +0000

Workaround crash freeing contour with cvc_next = -1

XXX: _WHY_ Does this occur??


---

diff --git a/src/polygon1.c b/src/polygon1.c
index cb7f62a..24e9dfa 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -3633,18 +3633,30 @@ poly_DelContour (PLINE ** c)
   for (cur = PREV_EDGE (&(*c)->head); cur != &(*c)->head; cur = prev)
     {
       prev = PREV_EDGE (cur);
-      if (cur->cvc_next != NULL)
-	{
-	  free (cur->cvc_next);
-	  free (cur->cvc_prev);
-	}
+
+      if (cur->cvc_next == -1)
+        fprintf (stderr, "WHY DID THIS HAPPEN?? cvc_next == -1 in poly_DelContour() !\n");
+      else
+        free (cur->cvc_next);
+
+      if (cur->cvc_prev == -1)
+        fprintf (stderr, "WHY DID THIS HAPPEN?? cvc_prev == -1 in poly_DelContour() !\n");
+      else
+        free (cur->cvc_prev);
+
       g_slice_free (VNODE, cur);
     }
-  if ((*c)->head.cvc_next != NULL)
-    {
-      free ((*c)->head.cvc_next);
-      free ((*c)->head.cvc_prev);
-    }
+
+  if ((*c)->head.cvc_next == -1)
+    fprintf (stderr, "WHY DID THIS HAPPEN?? cvc_next == -1 in poly_DelContour() !\n");
+  else
+    free ((*c)->head.cvc_next);
+
+  if ((*c)->head.cvc_prev == -1)
+    fprintf (stderr, "WHY DID THIS HAPPEN?? cvc_prev == -1 in poly_DelContour() !\n");
+  else
+    free ((*c)->head.cvc_prev);
+
   /* FIXME -- strict aliasing violation.  */
   if ((*c)->tree)
     {
